lib/msgServer/mmrp/index.js

```diff
@@ -103,11 +98,33 @@ function MmrpNode(role, cfg, clusterId) {
 	if (this.isRelay) {
 		this.router = createRouter(cfg);
 		var routerUri = this.router.getsockopt(zmq.ZMQ_LAST_ENDPOINT);
+
 		if (routerUri) {
 			this.routerPort = parseInt(routerUri.substr(routerUri.lastIndexOf(':') + 1), 10);
+
+			// Share MMRP master port through the environment
+			process.env[ENV_MAGE_MMRP_MASTER_PORT] = this.routerPort;
+
+			// Generate the cluster ID based on the port the master listens on
+			this.clusterId = serviceDiscovery.helpers.generateId(this.routerPort);
+
+			// Set the identity of the router socket to the cluster ID
+			this.router.setsockopt(zmq.ZMQ_IDENTITY, new Buffer(this.clusterId));
 		}
+	} else {
+		// Retrieve the cluster ID from the environment
+		const masterPort = process.env[ENV_MAGE_MMRP_MASTER_PORT];
+
+		// Generate the cluster ID based on the port the master listens on
+		this.clusterId = serviceDiscovery.helpers.generateId(masterPort);
 	}
 
+	// dealer
+
+	this.identity = this.isRelay ? this.clusterId : createDealerIdentity(this.clusterId);
```

Comment:
I get the impression from looking at the code that `this.clusterId` could be undefined. That is, there may be code paths that lead to it not getting set.

Reply:
I think its undefined if you are in single mode (when zmq.ZMQ_LAST_ENDPOINT on line 100 returns nothing).